security/pkg/server/ca/node_auth.go

```diff
@@ -0,0 +1,116 @@
+		// This should not happen; if it does, the CA will not function properly, and its likely a programming error, so returning
+		// an error here (which will ultimately exit the process) is appropriate.
+		return nil, fmt.Errorf("failed to add indexer: %v", err)
+	}
+	return &NodeAuthorizer{
+		podLister:           pods.Lister(),
+		podIndexer:          pods.Informer().GetIndexer(),
+		trustedNodeAccounts: trustedNodeAccounts,
+	}, nil
+}
+
+func (na *NodeAuthorizer) authenticateImpersonation(caller security.KubernetesInfo, requestedIdentityString string) error {
+	callerSa := types.NamespacedName{
+		Namespace: caller.PodNamespace,
+		Name:      caller.PodServiceAccount,
+	}
+	// First, make sure the caller is allowed to impersonate, in general
+	if _, f := na.trustedNodeAccounts[callerSa]; !f {
+		return fmt.Errorf("caller (%v) is not allowed to impersonate", caller)
+	}
+	// Next, make sure the identity they want to impersonate is valid, in general
+	requestedIdentity, err := spiffe.ParseIdentity(requestedIdentityString)
+	if err != nil {
+		return fmt.Errorf("failed to validate impersonated identity %v", requestedIdentityString)
+	}
+
+	// Finally, we validate the requested identity is running on the same node the caller is on
+	callerPod, err := na.podLister.Pods(caller.PodNamespace).Get(caller.PodName)
+	if err != nil {
+		return fmt.Errorf("failed to lookup pod: %v", err)
+	}
+	// Make sure UID is still valid for our current state
+	if callerPod.UID != types.UID(caller.PodUID) {
+		// This would only happen if a pod is re-created with the same name, and the CSR client is not in sync on which is current;
+		// this is fine and should be eventually consistent. Client is expected to retry in this case.
+		return fmt.Errorf("pod found, but UID does not match: %v vs %v", callerPod.UID, caller.PodUID)
+	}
+	if callerPod.Spec.ServiceAccountName != caller.PodServiceAccount {
+		// This should never happen, but just in case add an additional check
+		return fmt.Errorf("pod found, but ServiceAccount does not match: %v vs %v", callerPod.Spec.ServiceAccountName, caller.PodServiceAccount)
+	}
+	// We want to find out if there is any pod running with the requested identity on the callers node.
+	// The indexer (previously setup) creates a lookup table for a {Node, SA} pair, which we can lookup
+	k := SaNode{
+		ServiceAccount: types.NamespacedName{Name: requestedIdentity.ServiceAccount, Namespace: requestedIdentity.Namespace},
+		Node:           callerPod.Spec.NodeName,
+	}
+	// TODO: this is currently single cluster; we will need to take the cluster of the proxy into account
+	// to support multi-cluster properly.
+	res, err := na.podIndexer.ByIndex(NodeSaIndex, k.String())
+	if err != nil {
+		// This should never happen (only possible if the index doesn't exist)
+		return fmt.Errorf("failed to find node local service accounts: %v", err)
+	}
+	// We don't care what pods are part of the index, only that there is at least one. If there is one,
```

Comment:
Does this function checks that the ztunnel and the pod for which the ztunnel requests a certificate are on the same node?

Reply:
Exactly